astropy/coordinates/sky_coordinate.py

```diff
@@ -1850,79 +1850,70 @@ def radial_velocity_correction(
             # we do need space motion after all
             coo_at_rv_obstime = self.apply_space_motion(obstime)
-        elif self.obstime is None:
-            # warn the user if the object has differentials set
-            if "s" in self.data.differentials:
-                warnings.warn(
-                    "SkyCoord has space motion, and therefore the specified "
-                    "position of the SkyCoord may not be the same as "
-                    "the `obstime` for the radial velocity measurement. "
-                    "This may affect the rv correction at the order of km/s"
-                    "for very high proper motions sources. If you wish to "
-                    "apply space motion of the SkyCoord to correct for this"
-                    "the `obstime` attribute of the SkyCoord must be set",
-                    AstropyUserWarning,
-                )
+        elif self.obstime is None and "s" in self.data.differentials:
+            warnings.warn(
+                "SkyCoord has space motion, and therefore the specified "
+                "position of the SkyCoord may not be the same as "
+                "the `obstime` for the radial velocity measurement. "
+                "This may affect the rv correction at the order of km/s"
+                "for very high proper motions sources. If you wish to "
+                "apply space motion of the SkyCoord to correct for this"
+                "the `obstime` attribute of the SkyCoord must be set",
+                AstropyUserWarning,
+            )
 
-        pos_earth, v_earth = get_body_barycentric_posvel("earth", obstime)
-        if kind == "barycentric":
-            v_origin_to_earth = v_earth
-        elif kind == "heliocentric":
-            v_sun = get_body_barycentric_posvel("sun", obstime)[1]
-            v_origin_to_earth = v_earth - v_sun
-        else:
+        if kind not in ("barycentric", "heliocentric"):
             raise ValueError(
                 "`kind` argument to radial_velocity_correction must "
                 f"be 'barycentric' or 'heliocentric', but got '{kind}'"
             )
 
+        pos_earth, v_origin_to_earth = get_body_barycentric_posvel("earth", obstime)
+        if kind == "heliocentric":
+            v_origin_to_earth -= get_body_barycentric_posvel("sun", obstime)[1]
+
         gcrs_p, gcrs_v = location.get_gcrs_posvel(obstime)
         # transforming to GCRS is not the correct thing to do here, since we don't want to
         # include aberration (or light deflection)? Instead, only apply parallax if necessary
         icrs_cart = coo_at_rv_obstime.icrs.cartesian
-        icrs_cart_novel = icrs_cart.without_differentials()
-        if self.data.__class__ is UnitSphericalRepresentation:
-            targcart = icrs_cart_novel
-        else:
+        targcart = icrs_cart.without_differentials()
+        if self.data.__class__ is not UnitSphericalRepresentation:
             # skycoord has distances so apply parallax
```

Comment:
Here I'm more sure the code here was written to make clear what happens physically. I think we can keep that intent by rewriting the comment as
```
# SkyCoord has distances, so apply parallax by calculating
# the direction of the target as seen by the observer.

Reply:
OK, I'll do that.